[feature] 네이버 로그인을 통한 AccessToken 발급

jihye-woo opened the pull request:
#7 에서 카카오, 구글 로그인이 통합된 브랜치(feature/login)를 이어받아 네이버 로그인까지 통합 완료하였습니다.

jihye-woo committed:
[#4] feat : Naver 로그인 요청, 응답 DTO 구현

> NaverAccessTokenRequest

> NaverAccessTokenResponse

jihye-woo committed:
[#4] feat : 네이버 로그인 인터셉터 구현

> NaverOauthInterceptor
  - state 생성 후, session에 추가

> OauthConfig
  - NaverOauthInterceptor 인터셉터 추가
  - /login/naver url 요청에 한해 작동

jihye-woo committed:
[#4] feat : WebClient 빈 생성

jihye-woo committed:
[#4] feat : Naver 설정 정보 클래스 구현

> NaverOauthProperties 클래스 구현
  - oauth.naver라는 prefix를 가진 정보를 필드로 등록

jihye-woo committed:
[#4] feat : NaverOauthService 구현

> NaverOauthService::issueAccessToken 구현
  - webClient로 access token 요청 후 응답을 컨트롤러 계층으로 리턴

jihye-woo committed:
[#4] feat : ClientController 구현

> 로그인 시 클라이언트 역할 컨트롤러 구현
  - /login/naver 요청과 매핑
  - 생성된 code와 state 정보를 쿼리파라미터로 붙여 /api/login/naver로 리다이렉트

jihye-woo committed:
[#4] feat : 네이버 로그인 설정 정보 추가

> application-oauth-naver.properties 생성
  - 네이버 로그인에 필요한 정보
  - application.properties에 include 옵션으로 해당 프로파일 추가

jihye-woo committed:
[#4] feat : 네이버 로그인 서비스 로직 추가

> OauthServiceFactory에 네이버 로그인 로직 추가
  - dispatchNaverAccessTokenRequest()
  - 맵에서 빈 이름을 통해 빈 조회 후 서비스 로직 호출

> 파라미터로 code와 state
  - 발급된 엑세스토큰의 유효성을 확인하기 위해 부가적으로 state 토큰이 필요

jihye-woo committed:
[#4] feat : 네이버 로그인 요청 처리로직, 컨트롤러 계층에 추가

> LoginController::issueNaverAccessToken
  - 네이버 로그인 요청 url ("/api/login/naver")

jihye-woo committed:
[#4] refactor : component 패키지 삭제 및 파일 이동

> NaverOauthInterceptor
  - /component ➡️ /interceptor

> NaverOauthProperties
  - /component ➡️ /properties

> OauthConfig
  - /component ➡️ /config

jihye-woo committed:
[#4] fix : 중복 빈 webClient 제거

> component 아래 WebClientConfig 설정 클래스 삭제

jihye-woo committed:
[#4] fix : 중복 의존성 제거

- spring-boot-starter-webflux
- spring-boot-configuration-processor

jihye-woo commented on pintree/src/main/java/com/trio/pintree/login/service/NaverOauthService.java:
```diff
@@ -1,20 +1,37 @@
+    private final WebClient webClient;
 
     @Override
     public AccessTokenResponse issueAccessToken(String... str) {
```
저번 회의에서 이야기했던대로 가변인자로 받는 대신 code와 state 필드를 가진 클래스를 만들어서 인자로 사용하는게 더 나을 것 같아요!

sanhee replied:
가변인자는 의미상 파악하기도 어렵고, 자칫 잘못하면 `ArrayIndexOutOfBoundsException`가 쉽게 날 수 있다고 생각해서
저도 관련 클래스로 관리하는 것에 대해 동의합니다!

jihye-woo commented on pintree/src/main/java/com/trio/pintree/login/service/OauthServiceFactory.java:
```diff
@@ -22,4 +22,8 @@ public AccessTokenResponse dispatchGoogleAccessTokenRequest(String code) {
     public AccessTokenResponse dispatchKakaoAccessTokenRequest(String code) {
         return oauthServiceMap.get("kakaoOauthService").issueAccessToken(code);
     }
+
+    public AccessTokenResponse dispatchNaverAccessTokenRequest(String code, String state) {
+        return oauthServiceMap.get("naverOauthService").issueAccessToken(code, state);
+    }
```
일단 여기서는 앞선 코드의 형태를 맞춰 작성하긴 했지만,
개인적으로 생각하는 **팩토리 클래스의 역할**은 '**서비스 구현체들을 가지고 있다가 리턴해주는 것**'이라고 생각해요!!
그래서 지금처럼 팩토리 클래스의 메소드가 서비스 로직을 다 처리해서 결과를 반환해주기보다는 서비스 구현체들만 반환해주는 게 좋다고 생각했어요!

그렇게되면 각 클래스의 역할은 아래처럼 조금 더 명확해질 것 같습니다!
- `OauthServiceFactory` 클래스 -> 서비스 구현체 리턴
- `LoginService` 클래스 -> 로그인 서비스 로직 처리

sanhee replied:
듣고보니 그렇네요! 
이부분은 제가 리팩토링을 진행하다가 실수한 것 같은데, 우디가 말씀하신 대로 역할을 분리하는게 맞는 것 같아요!

jihye-woo reviewed (commented):
아래 토픽에 대해 간단하게 코멘트 달았습니다
1. 가변인자
2. 팩토리 클래스 역할

감사합니다 😊

sanhee commented on pintree/src/main/java/com/trio/pintree/login/controller/ClientController.java:
```diff
@@ -0,0 +1,46 @@
+        log.debug("Redirect URI: {}", redirectUri);
+        log.debug("State : {}", state);
+
+        return new StringBuilder("redirect:https://nid.naver.com/oauth2.0/authorize?")
+                .append("client_id=").append(clientId)
+                .append("&response_type=").append(responseType)
+                .append("&redirect_uri=").append(redirectUri)
+                .append("&state=").append(state)
+                .toString();
+    }
```
ClientController 명칭 마음에 드네요!!

또 scope value를 변수로 할당해주신점 휴먼 에러를 방지할 수 있어서 정말 마음에 듭니다! 
한 가지 마이너하게 우려되는 부분은 `client_id` 와 같은 `scope`가 추가됐을 때, 관리 포인트가 늘어난다는 점이 있을 수 있을 거 같아요!

sanhee commented on pintree/src/main/java/com/trio/pintree/login/dto/NaverAccessTokenRequest.java:
```diff
@@ -0,0 +1,39 @@
+import org.springframework.web.util.UriComponents;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Getter
+@RequiredArgsConstructor(staticName = "create")
+public class NaverAccessTokenRequest {
```
혹시 있을 사이드 이펙트 방지를 위해 Dto 클래스 필드 final로 지정해주신 거 정말 마음에 듭니다!!

그리고 `staticName` 옵션을 이용한 `static factory` 메소드 생성하는 꿀팁을 알게됐네요!! (줍줍🐸)

항상 꿀팁 알려주셔서 감사해요~

sanhee commented on pintree/src/main/java/com/trio/pintree/login/dto/NaverAccessTokenResponse.java:
```diff
@@ -0,0 +1,24 @@
+
+    @JsonProperty("expires_in")
+    private String expiresIn;
+
+}
```
엇 특별한 건 아니지만 나중에 공백하나 추가해야겠네요.

그러고 보니 인텔리제이에서 자동정렬 같은 IDE 스타일 정보를 export 할 수 있는데,
활용해서 팀 전용 스타일 만들어두면 편리할 것 같아요.

sanhee commented on pintree/src/main/java/com/trio/pintree/login/properties/NaverOauthProperties.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.web.util.UriComponents;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Data
+@Component
+@ConfigurationProperties(prefix = "oauth.naver")
+public class NaverOauthProperties {
```
롬복의 `@Data` 애노테이션은
 `@Getter / @Setter / @ToString / @EqualsAndHashCode / @RequiredArgsConstructor ` 를 전부 생성해주는 것으로 알고 있습니다!

혹시 Properties 클래스가 Equlas나 Tostring을 활용할 부분이 있을까요?
그게 아니라면, 불필요한 메서드도 함께 생성되는 것 같아, 필요한 메서드만 지정하는 게 직관적이고 좋을 것 같다는 개인적인 생각이 들었습니다!

sanhee replied:
그리고 마이너한 의견인데 
애노테이션 정의 순서도 컨벤션을 지정해서 일관성있게 맞추면 좋을 것 같습니다!

jihye-woo replied:
어노테이션 순서를 통일하는게 좋을 것 같네요! 좋은 의견 감사합니다!ㅎㅎ

너무 아무생각없이 사용했는데, 확실히 `@Data` 대신 필요한 기능만 달아놓는게 좋겠네요! 예리하시군요😎 구웃 감사합니다👏👏

sanhee commented on pintree/src/main/java/com/trio/pintree/login/properties/NaverOauthProperties.java:
```diff
@@ -0,0 +1,43 @@
+    private String grantType;
+    private String accessTokenUrl;
+
+    public NaverAccessTokenRequest tokenRequest(String code, String state){
+        return NaverAccessTokenRequest
+                .create(
+                        clientId,
+                        clientSecret,
+                        grantType,
+                        state,
+                        code
+                );
+    }
+
+    public String getAsUriParams(String code, String state){
+        UriComponents builder = UriComponentsBuilder.fromHttpUrl(accessTokenUrl)
+                .queryParam("client_id", clientId)
+                .queryParam("client_secret", clientSecret)
+                .queryParam("grant_type", grantType)
+                .queryParam("state", state)
+                .queryParam("code", code)
+                .build(false);
+
+        return builder.toString();
+    }
```
개인적으로 프로퍼티 클래스는 `VO`라고 인지하고 있었습니다.
이러한 관점에서 `tokenRequest()`, `getAsUriParams()` 메서드는 혹시 필요 이상의 역할을 떠안고 있는게 아닌가 생각해봤어요!

이부분은 제가 잘모르는 부분이라 편하게 같이 이야기 나눠보아요🐸

jihye-woo replied:
일단 두 메소드를 클래스 내부에 둔 이유는 클래스 내부의 값들이 모두 사용되어야 하는 로직이 들어있어서였어요!
메소드를 바깥으로 두게되면 Property 변수들을 모두 getter로 호출해서 사용해야하니까요

하지만 노을 말씀을 듣고 다시 생각해보니 유지보수 관점에서는 확실히 해당 로직들이 좀 더 연관된 곳으로 배치 되어야할 것 같다는 생각이 드네요!
### tokenRequest()
(물론 DTO와 VO는 다르긴 하지만), DTO 내부에서 toEntity()라는 메소드로 엔티티 클래스를 만들어주는 것과 비슷한 역할을 한다고 생각했습니다. 

그런데 NaverAccessTokenRequest를 만드는 로직은 NaverAccessTokenRequest 클래스 내부에 있어야할 것 같기도 하네요!
또 유지보수 관점에서 생각해보면 NaverAccessTokenRequest 생성자를 변경할 경우에 tokenRequest() 로직을 변경해야할 수도 있는데 **변경해야할 요소가 흩어져있으면** 안좋아보일 것 같네요!
(특히 String a, String b 와 같이 타입이 같은 필드 변수들 순서를 b, a로 변경하고나서 NaverOauthProperties 클래스의 메소드를 수정 해야할 생각은 전혀 못할 것 같아요)
만약 NaverAccessTokenRequest로 메소드 로직을 옮기게 된다면 아래와 같겠군요! 수정하는게 좋을 것 같아요! 감사합니다😄
```java

@Getter
@RequiredArgsConstructor(staticName = "create")
public class NaverAccessTokenRequest {
  // ...
 public NaverAccessTokenRequest tokenRequest(NaverOauthProperties properties, String code, String state){
        return new NaverAccessTokenRequest(
                properties.getClientId(),
                properties.getClientSecret(),
                properties.getGrantType(),
                state,
                code
        );
    }

```

### getAsUriParams()
이 메소드는 현재 webclient에 넣을 토큰 발행 GET 요청 url을 만들때 사용됩니다!(NaverOauthService 클래스)

노을이 말씀하셨던 VO 역할 관점에서 보면(아직 잘 모르긴 하지만요🥺)
범용적으로 사용되는 VO가 너무 구체적인 흐름(네이버 로그인시 엑세스 토큰을 얻을 때만 사용)에 의존적인 것 같네요! 이것도 수정하도록 하겠습니다아!ㅎㅎ 감사합니다😀

sanhee replied:
어제 스터디에서 `도메인 모델 패턴`과 `트랜잭션 스크립트 패턴` 을 학습하고 나니..약간 햇갈리는데...
좀 더 정리되면 말씀드리겠습니다!

jihye-woo replied:
저도 아직 정확하게는 모르지만 
도메인 모델 패턴과 같은 패턴은 도메인 객체에 적용되는게 맞고,
이렇게 변경되지 않는 프로퍼티 값들을 가진 프로퍼티 클래스는 VO에 가까운 것 같아서 성질이 사알짝 다른 것 같아요!

sanhee commented on pintree/src/main/resources/application-oauth-naver.properties:
```diff
@@ -0,0 +1,8 @@
+oauth.naver.responseType=code
+oauth.naver.grantType=authorization_code
+oauth.naver.redirectionUri=http://localhost:8080/api/login/naver
+oauth.naver.accessTokenUrl=https://nid.naver.com/oauth2.0/token
```
표지판🛑
(사실 리뷰 안남겨도 되지만 초창기 코드스쿼드 리뷰어분들 따라해봤습니다.)

sanhee commented on pintree/src/main/resources/application.properties:
```diff
@@ -1,3 +1,3 @@
 logging.level.com.trio.pintree=DEBUG
 
-spring.profiles.include=oauth-google,oauth-kakao
+spring.profiles.include=oauth-google,oauth-kakao,oauth-naver
```
개별 프로퍼티를 이런식으로 포함시켜서 관리하네요!
몰랐던 내용이라 배워갑니다!

sanhee commented on pintree/src/main/java/com/trio/pintree/login/dto/NaverAccessTokenRequest.java:
```diff
@@ -0,0 +1,39 @@
+
+    private final String code;
+
+    public String getAsUriParams(final String url){
+        UriComponents builder = UriComponentsBuilder.fromHttpUrl(url)
+                .queryParam("client_id", clientId)
+                .queryParam("client_secret", clientSecret)
+                .queryParam("grant_type", grantType)
+                .queryParam("state", state)
+                .queryParam("code", code)
+                .build(false);
+
+        return builder.toString();
+    }
```
해당 메서드가 파라미터만 추가된 채로 
동일 로직이 다른 클래스(properties..)에서도 정의 된 것 같은데

혹시 분리해서 작성하신 이유가 있을까요?

jihye-woo replied:
저의 오류를 또 매의 눈으로 찾아주셨군요! 감사합니다😄 
해당 메소드가 requestDTO 내부가 아니라 property 쪽에 있는게 재사용하기도 좋을 것 같아서 옮기게 되었는데,
그 과정에서 이전 메소드를 지우지 않았군요!ㅠ

사용하지 않는 메소드이므로 삭제하도록 하겠습니다!

sanhee reviewed (commented):
안녕하세요 우디 주말에 고생 많으셨습니다! 
리뷰를 하면서 생긴 궁금증과 개인적인 생각에 대해서 코멘트를 남겨드렸으니 
시간되실 때 확인해주시면 감사하겠습니다!

sanhee commented on pintree/src/main/java/com/trio/pintree/login/controller/ClientController.java:
```diff
@@ -0,0 +1,46 @@
+    private final String responseType;
+    private final String redirectUri;
+
+    public ClientController(@Value("${oauth.naver.clientId}") String clientId,
+                            @Value("${oauth.naver.responseType}") String responseType,
+                            @Value("${oauth.naver.redirectionUri}") String redirectUri) throws NoSuchAlgorithmException {
```
혹시 이부분은 다른 클래스와 다르게 `@ConfigurationProperties` 와 같은 애노테이션을 사용하지 않은 이유가 있으신가요? 이점이라던가 궁금해서 여쭤봅니다!

jihye-woo replied:
원래 코드들도 `@Value`를 사용하다가 리펙토링을 진행하면서 `@ConfigurationProperties`를 도입하게되었는데요, 처음에는 `ClientController` 클래스가 이후에 없어질 코드라고 생각해서 따로 리펙토링을 진행하진 않았습니다🥺

그런데 다시 생각해보니 아무리 테스트용이라도 메인 코드들과 연관성이 있는 코드라면 함께 리펙토링해주는게 좋겠네요!🤔

jihye-woo reviewed (commented):
세세하게 리뷰해주셔서 감사합니다 @sanhee 😄
그에 비해서 코멘트가 너무 늦었네요ㅠㅠ

편한 시간에 확인해주시면 감사드리겠습니다!

jihye-woo committed:
[#8] refactor : 사용하지 않는 NaverAccessTokenRequest 클래스 삭제

> webclient 적용 후 사용하지 않음

jihye-woo committed:
[#8] refactor : 코드 마지막 줄 한 칸 띄우기

jihye-woo committed:
[#8] refactor : NaverOauthProperties 메소드 삭제

> tokenRequest()
  - 현재 사용되지 않는 메소드
  - NaverAccessTokenRequest 클래스가 있다면, 해당 클래스 아래로 이동하는 것이 더 바람직

> getAsUriParams()
  - 네이버 로그인시 수행되는 access token 발급 요청 로직을 수행하는 곳으로 이동하는 것이 바람직

sanhee reviewed (approved):
고생 많으셨습니다!!
